[BugFix]Avoid printing token in be log

xyllq999 opened the pull request:
## Why I'm doing:

## What I'm doing:

Fixes #issue

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [x] This pr needs user documentation (for new or modified features or behaviors)
  - [x] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.3
  - [x] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

xyllq999 commented:
> #### [FE Incremental Coverage Report]
> ✅ **pass** : 0 / 0 (0%)

@kevincai Could you please review this pr?

gengjun-git reviewed (changes requested):
Better to ignore the token log it in the upper layer call.

xyllq999 commented:
> Better to ignore the token log it in the upper layer call.

The upper layer is a lot of places where the TMasterInfo object is printed, and the < < operator is overloaded. The actual print is defined in the printTo function.

kevincai commented:
> > #### [FE Incremental Coverage Report]
> > ✅ **pass** : 0 / 0 (0%)
> 
> @kevincai Could you please review this pr?

I am hesitated to agree processing the sensitive info by diving into thrift library. Think about protobuf library, it could have the similar issue.

And because it is so deep in the library, the affection of this change is hard to be evaluated.

xyllq999 commented:
> > > #### [FE Incremental Coverage Report]
> > > ✅ **pass** : 0 / 0 (0%)
> > 
> > 
> > @kevincai Could you please review this pr?
> 
> I am hesitated to agree processing the sensitive info by diving into thrift library. Think about protobuf library, it could have the similar issue.
> 
> And because it is so deep in the library, the affection of this change is hard to be evaluated.

This function printTo is used only for the < < overloaded operator of the TMasterInfo object.

xyllq999 commented:
> > > > > #### [FE Incremental Coverage Report]
> > > > > ✅ **pass** : 0 / 0 (0%)
> > > > 
> > > > 
> > > > @kevincai Could you please review this pr?
> > > 
> > > 
> > > I am hesitated to agree processing the sensitive info by diving into thrift library. Think about protobuf library, it could have the similar issue.
> > > And because it is so deep in the library, the affection of this change is hard to be evaluated.
> > 
> > 
> > This function printTo is used only for the < < overloaded operator of the TMasterInfo object.
> 
> 1. overload or patching the `operator<<` to limit the impact
> 2. or whenever there is a sensitive info leaking, remember to copy a TMasterInfo object and reset its token before printing.

operator<< is also in thrift code，you can not change it directly

xyllq999 commented:
> > > > > #### [FE Incremental Coverage Report]
> > > > > ✅ **pass** : 0 / 0 (0%)
> > > > 
> > > > 
> > > > @kevincai Could you please review this pr?
> > > 
> > > 
> > > I am hesitated to agree processing the sensitive info by diving into thrift library. Think about protobuf library, it could have the similar issue.
> > > And because it is so deep in the library, the affection of this change is hard to be evaluated.
> > 
> > 
> > This function printTo is used only for the < < overloaded operator of the TMasterInfo object.
> 
> 1. overload or patching the `operator<<` to limit the impact
> 2. or whenever there is a sensitive info leaking, remember to copy a TMasterInfo object and reset its token before printing.

![image](https://github.com/user-attachments/assets/d2c1904c-37a7-4c72-87c2-b7502dfa62ef)

kevincai commented:
Those patches are either for perf purpose or for compiling fix. None of them are changing the underlying behavior directly.

xyllq999 commented:
> Those patches are either for perf purpose or for compiling fix. None of them are changing the underlying behavior directly.



> Those patches are either for perf purpose or for compiling fix. None of them are changing the underlying behavior directly.

Okay, I see what you mean. Try not to modify the source code behavior. I'll re-produce a PR of the masking function where it is used.

xyllq999 committed:
避免BE日志打印铭文token

Signed-off-by: xuyan103@huawei.com
Signed-off-by: xyllq999 <734432375@qq.com>

xyllq999 committed:
避免BE日志打印铭文token

Signed-off-by: xuyan103@huawei.com
Signed-off-by: xyllq999 <734432375@qq.com>

xyllq999 committed:
避免BE日志打印铭文token

Signed-off-by: xuyan103@huawei.com
Signed-off-by: xyllq999 <734432375@qq.com>

xyllq999 committed:
Merge branch 'main' of https://github.com/StarRocks/starrocks into main-token

xyllq999 committed:
避免BE日志打印铭文token

Signed-off-by: xuyan103@huawei.com
Signed-off-by: xyllq999 <734432375@qq.com>

xyllq999 committed:
避免BE日志打印铭文token

Signed-off-by: xuyan103@huawei.com
Signed-off-by: xyllq999 <734432375@qq.com>

xyllq999 commented:
> Better to ignore the token log it in the upper layer call.

Please check this PR again, I have modified the implementation and added UT, but it seems that there is something wrong with the UT check of BE, please help solve it.

kevincai commented on be/src/agent/heartbeat_server.cpp:
```diff
@@ -93,18 +95,19 @@ void HeartbeatServer::heartbeat(THeartbeatResult& heartbeat_result, const TMaste
                                "with a larger epoch";
     } else if (*res == kNeedUpdateAndReport) {
-        LOG(INFO) << "Updating master info: " << master_info;
+        LOG(INFO) << "Updating master info: " << print_master_info(master_info);
```
```suggestion
        MasterInfo info(master_info)
        if (info.__isset.token) {
            info.__set_token = "<hidden>";
        }
        LOG(INFO) << "Updating master info: " << master_info;
```

kevincai commented on be/src/agent/heartbeat_server.cpp:
```diff
@@ -150,6 +153,50 @@ void HeartbeatServer::heartbeat(THeartbeatResult& heartbeat_result, const TMaste
 }
 
+std::string HeartbeatServer::print_master_info(const TMasterInfo& master_info) const {
+    using ::apache::thrift::to_string;
```
or here make a copy of master_info, clear the token and then use thrift generated TMasterInfo::printTo() to generate the string.

xyllq999 committed:
避免BE日志打印铭文token

Signed-off-by: xuyan103@huawei.com
Signed-off-by: xyllq999 <734432375@qq.com>

xyllq999 committed:
避免BE日志打印明文token

Signed-off-by: xuyan103@huawei.com
Signed-off-by: xyllq999 <734432375@qq.com>

kevincai commented on be/src/agent/heartbeat_server.h:
```diff
@@ -79,6 +79,9 @@ class HeartbeatServer : public HeartbeatServiceIf {
     StatusOr<CmpResult> compare_master_info(const TMasterInfo& master_info);
 
+    // New function to print TMasterInfo object as a string
+    std::string print_master_info(const TMasterInfo& master_info) const;
```
this interface is not necessary to be part of the HeartbeatServer, for just because HeartbeatServer uses the interface to print the master info.

Try to find a better place for this interface, e.g. common/ or util/ ?

xyllq999 committed:
避免BE日志打印明文token

Signed-off-by: xuyan103@huawei.com
Signed-off-by: xyllq999 <734432375@qq.com>

kevincai commented on be/src/agent/heartbeat_server.cpp:
```diff
@@ -150,6 +152,20 @@ void HeartbeatServer::heartbeat(THeartbeatResult& heartbeat_result, const TMaste
+
+    master_info_copy.printTo(out);
+
+    return out.str();
```
```suggestion
    std::ostringstream out;
    if (!master_info.__isset.token) {
        master_info.printTo(out);
    } else {
        MasterInfo info(master_info);
        info.__set_token = "<hidden>";
        info.printTo(out);
    }
    return out.str();
```

Save the object clone if the token is not set.

xyllq999 committed:
避免BE日志打印明文token

Signed-off-by: xuyan103@huawei.com
Signed-off-by: xyllq999 <734432375@qq.com>

xyllq999 committed:
避免BE日志打印明文token

Signed-off-by: xuyan103@huawei.com
Signed-off-by: xyllq999 <734432375@qq.com>

xyllq999 committed:
避免BE日志打印明文token

Signed-off-by: xuyan103@huawei.com
Signed-off-by: xyllq999 <734432375@qq.com>

xyllq999 committed:
避免BE日志打印明文token

Signed-off-by: xuyan103@huawei.com
Signed-off-by: xyllq999 <734432375@qq.com>

xyllq999 committed:
避免BE日志打印明文token

Signed-off-by: xuyan103@huawei.com
Signed-off-by: xyllq999 <734432375@qq.com>

xyllq999 committed:
避免BE日志打印明文token

Signed-off-by: xuyan103@huawei.com
Signed-off-by: xyllq999 <734432375@qq.com>

xyllq999 committed:
避免BE日志打印明文token

Signed-off-by: xuyan103@huawei.com
Signed-off-by: xyllq999 <734432375@qq.com>

xyllq999 committed:
避免BE日志打印明文token

Signed-off-by: xuyan103@huawei.com
Signed-off-by: xyllq999 <734432375@qq.com>

xyllq999 commented:
@kevincai I saw some sql use cases fail, but I didn't change that part? Can you help me see what the problem is? If there is no problem, can you help to restart the task?

xyllq999 commented:
> Better to ignore the token log it in the upper layer call.

I have modified the code and added UT, could you please help check the code?

wangsimo0 commented:
ignore backport check: 3.3.6
